kvcoord: optionally update range cache with descriptor returned in NLHE

arulajmani opened the pull request:
We recently started shipping the range descriptors back on
NotLeaseHolderError. Prior to this patch, we only ever used their
generation to elide certain lease updates if they originated from a
replica that had an older (stale) view of this world.

This patch goes a step further in making use of the returned range
descriptor -- we now update the client's range cache if the returned
range descriptor is newer than what existed on the client. We do this
by picking the freshest range descriptor and lease, independently, from
the range descriptor/lease between what exists in the client's range
cache and what was returned in the NLHE. Though unlikely, independently
choosing the freshest lease/range descriptor can lead to cases where the
lease is not compatible with the range descriptor. If we detect this to
be the case, we empty out the lease and simply cache the freshest range
descriptor.

Previously, we always accepted speculative leases to be more recent than
anything already in the cache. Now, we discard speculative leases if
they're coming from a replica that has an older view of the world.
Morally, these semantics are the right way to conceptualize https://github.com/cockroachdb/cockroach/issues/72772,
given that hazard could only ever exist for speculative leases. https://github.com/cockroachdb/cockroach/issues/82802
also falls out as a special case of this conceptualization, so closing
that issue simply entails adding a regression test.

We also get rid of optimizations where we would try to identify "stale"
range descriptors when updating lease information. These no longer make
sense given we don't just update the lease, we also update the range
descriptor. These optimizations also didn't account for all possible
cases, such as when the leaseholder was present on the range
descriptor, but as a LEARNER. See https://github.com/cockroachdb/cockroach/issues/75742 for more details about how
this hazard manifests.

With this new approach we also address https://github.com/cockroachdb/cockroach/issues/82802. Instead of invalidating
the routing when a descriptor is identified as stale and using that as
a proxy to bail early when routing to replicas, we instead switch to a
more direct approach -- if at any point we detect the leaseholder isn't
included on the transport, we exit the routing logic early, and retry
at a layer above. Given our new range cache update semantics using
descriptors on NLHE errors, it's quite likely this retry circumvents a
range descriptor lookup -- instead, we'd expect this to simply amount to
trying again with a newly constructed transport. The rationale is that
even though the client may have an arbitrarily stale view of the range,
we never expect it to regress. Thus, if the leaseholder was updated on
the client and it doesn't exist on the transport, it must be the case
that the transport was constructed using a stale range descriptor and
stale leaseholder information. As such, there is likely not much value
in trying to exhaust the transport (and potentially getting caught in
backoff loops). We're much better served by bailing early and trying
again with a fresh transport.

Fixes https://github.com/cockroachdb/cockroach/issues/82802
Fixes https://github.com/cockroachdb/cockroach/issues/75742

Release note: None
Release justification: Fixes for high priority bugs in existing
functionality.

arulajmani committed:
kvcoord: improve TestDistSenderIgnoresNLHEBasedOnOldRangeGeneration

Previously, this test was constructing a NLHE with a structure we
wouldn't really expect. We instead move to running two versions of
this test -- one with a speculative lease and another where the lease
returned has an older lease sequence than what is cached.

Release note: None
Release justification: Non-production code change

arulajmani committed:
kvclient: improve logging around range descriptor eviction

Previously, we were logging that we had evicted a range descriptor
from the cache even though we may not have -- we only evict if the
cache doesn't have a newer descriptor than the one we're trying to
evict. We now only log if we indeed evicted the range descriptor;
otherwise, we do not.

Release note: None
Release justification: Low risk, high benefit change

arulajmani committed:
kvcoord: optionally update range cache with descriptor returned in NLHE

We recently started shipping the range descriptors back on
NotLeaseHolderError. Prior to this patch, we only ever used their
generation to elide certain lease updates if they originated from a
replica that had an older (stale) view of this world.

This patch goes a step further in making use of the returned range
descriptor -- we now update the client's range cache if the returned
range descriptor is newer than what existed on the client. We do this
by picking the freshest range descriptor and lease, independently, from
the range descriptor/lease what exist in the client's range cache and
what was returned in the NLHE. Though unlikely, independently choosing
the freshest lease/range descriptor can lead to cases where the lease is
not compatible with the range descriptor. If we detect this to be the
case, we empty out the lease and simply cache the freshest range
descriptor.

Previously, we always accepted speculative leases to be more recent than
anything already in the cache. Now, we discard speculative leases if
they're coming from a replica that has an older view of the world.
Morally, these semantics are the right way to conceptualize #72772,
given that hazard could only ever exist for speculative leases. #82802
also falls out as a special case of this conceptualization, and this
patch adds a regression test for the scenario where an uninitialized
replica returns a NLHE with a speculative lease pointing to a replica
that isn't part of the range.

We also get rid of optimizations where we would try to identify "stale"
range descriptors when updating lease information. These no longer make
sense given we don't just update the lease, we also update the range
descriptor. These optimizations also didn't account for all possible
cases, such as when the leaseholder was present on the range
descriptor, but as a LEARNER. See #75742 for more details about how
this hazard manifests.

With this new approach we also address #75742. Instead of invalidating
the routing when a descriptor is identified as stale and using that as
a proxy to bail early when routing to replicas, we instead switch to a
more direct approach -- if at any point we detect the leaseholder isn't
included on the transport, we exit the routing logic early, and retry
at a layer above. Given our new range cache update semantics using
descriptors on NLHE errors, it's quite likely this retry circumvents a
range descriptor lookup from meta2 -- instead, we'd expect this to
simply amount to trying again with a newly constructed transport. The
rationale is that even though the client may have an arbitrarily stale
view of the range, we never expect it to regress. Thus, if the
leaseholder was updated on the client and it doesn't exist on the
transport, it must be the case that the transport was constructed using
a stale range descriptor and stale leaseholder information. As such,
there is likely not much value in trying to exhaust the transport
(and potentially getting caught in backoff loops). We're much better
served by bailing early and trying again with a fresh transport.

Fixes #82802
Fixes #75742

Release note: None
Release justification: Fixes for high priority bugs in existing
functionality.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/85503)
<!-- Reviewable:end -->

ajwerner commented on pkg/startupmigrations/migrations.go:
```diff
@@ -626,7 +626,7 @@ func (m *Manager) EnsureMigrations(ctx context.Context, bootstrapVersion roachpb
 		}
 
 		if log.V(1) {
-			log.Infof(ctx, "running migration %q", migration.name)
```
why?

arulajmani replied:
Wasn't intentional, reverted.

erikgrinaker reviewed (approved):
This looks great (modulo build failures), thanks for fixing this!

> Reviewed 1 of 1 files at r1, 1 of 1 files at r2, 7 of 7 files at r3, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/85503#-:-N8Y77T8B0FNUnkhx-c0:b46k5mi)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @aayushshah15 and @nvanbenschoten)



<!-- Sent from Reviewable.io -->

arulajmani reviewed (commented):
This was missing the case where the descriptor returned in the NLHE was incompatible with what was in the token/cache, which was the reason for a fair few build failures. We now account for this case by:
- treating things as a no-op if the incompatible descriptor is older than what's in the cache. 
- invalidating the token and updating the cache with the supplied descriptor/lease pair if the descriptor is newer. 

Separately, @nvanbenschoten, after our discussion yesterday I realized that the way I've structured this code we'll never take advantage of how uninitialized replicas construct their NotLeaseHolderErrrors (https://github.com/cockroachdb/cockroach/blob/87da966cbc15e058591d80c477c78d6059483a5c/pkg/kv/kvserver/store_send.go#L176-L194 being the code snippet from the original issue). As is, even if the replica returned in the NLHE error is part of the transport, we'll try the next replica on the transport instead of the one returned by an uninitialized replica. I'm a bit partial to keeping things the way they are as it keeps things understandable, but we could add some special casing to bring this behaviour back. Do you think this optimization is worth it to do so?

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/85503#-:-N8ddNuK09-H9F3aLWay:bj2rud0)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @aayushshah15, @erikgrinaker, and @nvanbenschoten)



<!-- Sent from Reviewable.io -->

nvb reviewed (commented):
I'm still working my way through the tests, but flushing what I have so far.

> Reviewed 1 of 1 files at r1, 1 of 1 files at r2, 1 of 7 files at r3, 1 of 4 files at r4, 1 of 3 files at r5, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/85503#-:-N8eQyMbCn8fBbbTeU-4:b-ij6wa2)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @aayushshah15, @arulajmani, and @erikgrinaker)

___
*[`-- commits` line 4 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/85503#-N8Zk4K0E63sp0m8ZlU8):*
Reminder to address this TODO.
___
*[`-- commits` line 36 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/85503#-N8dmtIH5k_kuUAk0_P_):*
TODO: is this still true?
___
*[`pkg/kv/kvclient/rangecache/range_cache.go` line 317 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/85503#-N8dnaMpCf8nguRr7Qsc:-N8dnaMpCf8nguRr7Qsd:bkhemu6) ([raw file](https://github.com/cockroachdb/cockroach/blob/9230f9aaef000ded75efa63258a183128ae51a75/pkg/kv/kvclient/rangecache/range_cache.go#L317)):*
> ```Go
> // information, returns nil. The result is considered immutable.
> func (et EvictionToken) Lease() *roachpb.Lease {
> 	return et.lease
> ```

nit: should this have a `if !et.Valid() {` check like we have in `Desc()`?
___
*[`pkg/kv/kvclient/rangecache/range_cache.go` line 367 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/85503#-N8e4r1I-URoFe0R7npy:-N8e4r1JCFoo47K-IvRF:b-r46ocu) ([raw file](https://github.com/cockroachdb/cockroach/blob/9230f9aaef000ded75efa63258a183128ae51a75/pkg/kv/kvclient/rangecache/range_cache.go#L367)):*
> ```Go
> // EvictionToken itself).
> //
> // The bool `updatedLeaseholder` is true if the leaseholder was updated in the
> ```

nit: this makes it sounds like a parameter. Consider `The returned bool ...`.
___
*[`pkg/kv/kvclient/rangecache/range_cache.go` line 368 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/85503#-N8e5Jyc0DtrjWA1meSa:-N8e5Jyc0DtrjWA1meSb:bxhekv1) ([raw file](https://github.com/cockroachdb/cockroach/blob/9230f9aaef000ded75efa63258a183128ae51a75/pkg/kv/kvclient/rangecache/range_cache.go#L368)):*
> ```Go
> //
> // The bool `updatedLeaseholder` is true if the leaseholder was updated in the
> // cache (i.e the passed-in lease was more recent than the cached lease).
> ```

What about if the cache had a lease that was more recent than that in both the token and the one which was handed in? We will update the token in that case, right? Will we return true?
___
*[`pkg/kv/kvclient/rangecache/range_cache.go` line 381 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/85503#-N8eLbKyCsSVtLmmAklQ:-N8eLbKyCsSVtLmmAklR:brv50ld) ([raw file](https://github.com/cockroachdb/cockroach/blob/9230f9aaef000ded75efa63258a183128ae51a75/pkg/kv/kvclient/rangecache/range_cache.go#L381)):*
> ```Go
> // in-turn will be overwritten by any subsequent update).
> func (et *EvictionToken) SyncTokenAndMaybeUpdateCache(
> 	ctx context.Context, l *roachpb.Lease, rangeDesc *roachpb.RangeDescriptor,
> ```

Is the range descriptor guaranteed to be non-empty on every error path? Even in mixed version clusters?
___
*[`pkg/kv/kvclient/rangecache/range_cache.go` line 387 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/85503#-N8e59sBBCEypVooG1yU:-N8e59sBBCEypVooG1yV:b-943me9) ([raw file](https://github.com/cockroachdb/cockroach/blob/9230f9aaef000ded75efa63258a183128ae51a75/pkg/kv/kvclient/rangecache/range_cache.go#L387)):*
> ```Go
> 	defer rdc.rangeCache.Unlock()
> 
> 	// TODO(arul): Check for compatibility between the supplied range descriptor
> ```

You were planning to address this TODO in this PR, right?
___
*[`pkg/kv/kvclient/rangecache/range_cache.go` line 391 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/85503#-N8eQXcpBK3jGP0gpgYm:-N8eQXcpBK3jGP0gpgYn:byna72g) ([raw file](https://github.com/cockroachdb/cockroach/blob/9230f9aaef000ded75efa63258a183128ae51a75/pkg/kv/kvclient/rangecache/range_cache.go#L391)):*
> ```Go
> 	// maybeUpdate does) if the descriptors are compatible.
> 
> 	stillValid, cachedEntry, rawEntry := et.syncRLocked(ctx)
> ```

How should this method behave when this sync updates the leaseholder but the `maybeUpdate` below does not? Should it return `true`?
___
*[`pkg/kv/kvclient/rangecache/range_cache.go` line 410 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/85503#-N8eLqVm-5oBOm8SoSCJ:-N8eLqVm-5oBOm8SoSCK:behpw3v) ([raw file](https://github.com/cockroachdb/cockroach/blob/9230f9aaef000ded75efa63258a183128ae51a75/pkg/kv/kvclient/rangecache/range_cache.go#L410)):*
> !descsCompatible(rangeDesc, et.Desc())

We're checking this condition twice. Should we rewrite as:
```go
if !descsCompatible(rangeDesc, et.Desc()) {
    if rangeDesc.Generation < et.Desc().Generation {
        return false
    }
    ...
}
```
___
*[`pkg/kv/kvclient/rangecache/range_cache.go` line 418 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/85503#-N8eMvELEKZGJLPvxH42:-N8eMvEM5sGKuFbt56_K:b-sxnl2i) ([raw file](https://github.com/cockroachdb/cockroach/blob/9230f9aaef000ded75efa63258a183128ae51a75/pkg/kv/kvclient/rangecache/range_cache.go#L418)):*
> ```Go
> 		// The cachedEntry wasn't updated; no need to replace it with newEntry in
> 		// the RangeCache.
> 		return updatedLeaseholder
> ```

If `!updated`, don't we know that `!updatedLeaseholder`? If so, then should we `return false`?
___
*[`pkg/kv/kvclient/rangecache/range_cache.go` line 426 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/85503#-N8ePyxTAwzJxKR02VEO:-N8ePyxTAwzJxKR02VEP:b-2bc2ul) ([raw file](https://github.com/cockroachdb/cockroach/blob/9230f9aaef000ded75efa63258a183128ae51a75/pkg/kv/kvclient/rangecache/range_cache.go#L426)):*
> ```Go
> 	et.desc = newEntry.Desc()
> 	et.lease = newEntry.leaseEvenIfSpeculative()
> 	return updatedLeaseholder
> ```

Are we ok with returning false here if we updated the descriptor but not the leaseholder? Will the client handle that case correctly?
___
*[`pkg/kv/kvclient/rangecache/range_cache.go` line 1338 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/85503#-N8eMEcM2VkUXT_wXXfn:-N8eMEcM2VkUXT_wXXfo:b-pyco7m) ([raw file](https://github.com/cockroachdb/cockroach/blob/9230f9aaef000ded75efa63258a183128ae51a75/pkg/kv/kvclient/rangecache/range_cache.go#L1338)):*
> ```Go
> // descriptor by comparing the receiver's fields to the passed-in parameters.
> // The updatedLease retval indicates whether the passed-in lease appears to be
> // newer than the lease the entry had before.
> ```

Do you want to give this paragraph a similar "If updatedLease is returned true, ..." sentence?
___
*[`pkg/kv/kvclient/rangecache/range_cache.go` line 1340 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/85503#-N8eMfhzA8dNelUWLauL:-N8eMfhzA8dNelUWLauM:b-whoqvz) ([raw file](https://github.com/cockroachdb/cockroach/blob/9230f9aaef000ded75efa63258a183128ae51a75/pkg/kv/kvclient/rangecache/range_cache.go#L1340)):*
> ```Go
> // newer than the lease the entry had before.
> //
> // The updated retval indicates if either the passed-in lease or the range
> ```

I found the combination of an `updatedLease` and an `updated` return value to be confusing. It wasn't clear from the caller whether one implied the other or why I needed to care about both. It would be good to detail that.

It's minor, but I think switching the order of these return values would help. I say that because if `updated` is a broader condition than `updatedLease`, then most callers will first consult `updated` and then only care about `updatedLease` is `updated` is true.
___
*[`pkg/kv/kvclient/rangecache/range_cache.go` line 1357 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/85503#-N8eMbYJ2Emmbw35Ozc3:-N8eMbYJ2Emmbw35Ozc4:belaz8p) ([raw file](https://github.com/cockroachdb/cockroach/blob/9230f9aaef000ded75efa63258a183128ae51a75/pkg/kv/kvclient/rangecache/range_cache.go#L1357)):*
> ```Go
> func (e *CacheEntry) maybeUpdate(
> 	ctx context.Context, l *roachpb.Lease, rangeDesc *roachpb.RangeDescriptor,
> ) (updatedLease bool, updated bool, newEntry *CacheEntry) {
> ```

nit: `) (updatedLease, updated bool, newEntry *CacheEntry) {`
___
*[`pkg/kv/kvclient/rangecache/range_cache.go` line 1363 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/85503#-N8eOR_ODe8pT5zv3vnP:-N8eOR_ODe8pT5zv3vnQ:bqwsrzj) ([raw file](https://github.com/cockroachdb/cockroach/blob/9230f9aaef000ded75efa63258a183128ae51a75/pkg/kv/kvclient/rangecache/range_cache.go#L1363)):*
> ```Go
> 	}
> 
> 	newEntry = &CacheEntry{
> ```

I feel like this method would be easier to reason about if you initially cloned the receiver and then updated to lease and desc, instead of working backwards like you're doing here (assuming both will be updated unless proven otherwise). Was there are reason to write it like this?
___
*[`pkg/kv/kvclient/rangecache/range_cache.go` line 1387 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/85503#-N8eNieJ0voWUDM6afB6:-N8eNieJ0voWUDM6afB7:b28qegw) ([raw file](https://github.com/cockroachdb/cockroach/blob/9230f9aaef000ded75efa63258a183128ae51a75/pkg/kv/kvclient/rangecache/range_cache.go#L1387)):*
> l.Equal(e.Lease())

I know if was here before, but this is suspect. It's comparing every field in the lease, which I'm not sure we ever want to do on the client. Is it needed to determine that two speculative leases are identical? If not, are there cases where the lease sequence is insufficient to distinguish the equality of leases?
___
*[`pkg/kv/kvclient/rangecache/range_cache.go` line 1411 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/85503#-N8eOwfC2ZjBY0kvRBx6:-N8eOwfC2ZjBY0kvRBx7:b-wbz1ru) ([raw file](https://github.com/cockroachdb/cockroach/blob/9230f9aaef000ded75efa63258a183128ae51a75/pkg/kv/kvclient/rangecache/range_cache.go#L1411)):*
> ```Go
> 	// down seems like the right thing to do, do we expect this to happen in the
> 	// wild?
> 	_, ok := newEntry.Desc().GetReplicaDescriptorByID(newEntry.lease.Replica.ReplicaID)
> ```

nit: why `.Desc()` and not `.desc`?


<!-- Sent from Reviewable.io -->

arulajmani committed:
kvserver: stop returning speculative leases from uninitialized replicas

Previously, if a request was routed to an uninitialized replica, the
returned NLHE would contain a speculative lease  pointing to the creator
of the uninitialized replica.

The prior commit makes it such that lease/leaseholder information in
NLHEs is only considered if accompanied by a non-stale range descriptor.
As uninitialized replicas always return empty range descriptors, which
are considered stale, the speculative lease they returned would never
be acted upon. Given this, we omit speculative leases in NLHEs returned
by uninitialized replicas. The idea being the client will simply move
on to the next replica in its transport which will very likely have
more useful leaseholder information for the client.

Release note: None
Release justification: Low risk, high benefit change

arulajmani reviewed (commented):
RFAL. 

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/85503#-:-N9HxYQm8dlxKUdPcAbu:by76qvq)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @aayushshah15, @erikgrinaker, and @nvanbenschoten)

___
*[`-- commits` line 4 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/85503#-N8Zk4K0E63sp0m8ZlU8):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

Reminder to address this TODO.
</blockquote></details>

Oops, done. 
___
*[`-- commits` line 36 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/85503#-N8dmtIH5k_kuUAk0_P_):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

TODO: is this still true?
</blockquote></details>

I _think_ the only scenario this is possible is the following scenario:

The cache had no lease information to begin with. It receives a NLHE from a replica that has both a stale range descriptor + lease information. In such a scenario, as written, we choose to insert the lease into the cache. It's possible the stale lease and cached range descriptor are incompatible. 

In my recent round of changes, I chose to structure things such that we don't compare range descriptor generations if the cached lease is empty. The idea being that a stale range descriptor doesn't necessarily imply stale lease information, so in the absence of anything better, let's just try it. How do you feel about this? 
___
*[`pkg/kv/kvclient/rangecache/range_cache.go` line 368 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/85503#-N8e5Jyc0DtrjWA1meSa:-N9HprAPAt6YmwEmzSo8:bh5l2u5) ([raw file](https://github.com/cockroachdb/cockroach/blob/9230f9aaef000ded75efa63258a183128ae51a75/pkg/kv/kvclient/rangecache/range_cache.go#L368)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

What about if the cache had a lease that was more recent than that in both the token and the one which was handed in? We will update the token in that case, right? Will we return true?
</blockquote></details>

We'll update the token but we'll return false. This behaviour is unchanged from what existed before. However, looking at why we use this information, I don't think the caller necessarily wants to make this distinction. 

Like we discussed offline, much of this can be simplified by changing the conditional at the caller to be `lh.IsSame(curReplica)` and we can clean up the amount of state that is tracked here and passed between callers. I've added a TODO to this effect, but I'll probably get to it in a quick followup patch if that's okay. 
___
*[`pkg/kv/kvclient/rangecache/range_cache.go` line 381 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/85503#-N8eLbKyCsSVtLmmAklQ:-N9CmyxkAQbbSG-QwCQR:b-o2u7ua) ([raw file](https://github.com/cockroachdb/cockroach/blob/9230f9aaef000ded75efa63258a183128ae51a75/pkg/kv/kvclient/rangecache/range_cache.go#L381)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

Is the range descriptor guaranteed to be non-empty on every error path? Even in mixed version clusters?
</blockquote></details>

This wasn't the behaviour in the patch you reviewed, but given our discussion on slack to change NLHEs returned from uninitialized replicas to not include a speculative lease, I'll describe how things look with the patch I'm going to push out. The only time we can have a non-empty range descriptor here is if we're running a 22.(1,2) mixed version cluster and an uninitialized replica on the 22.1 node returns a NLHE. That should be okay though, as we'll fail the range descriptor compatibility check below and no-op given the range descriptor's generation is 0. Further up the stack, the client will move on and try the next replica on its transport. I've modified the `TestDistSenderNLHEFromUninitializedReplicaDoesNotCauseUnboundedBackoff` test to simulate both the mixed version and the 22.2 version scenarios. 
___
*[`pkg/kv/kvclient/rangecache/range_cache.go` line 387 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/85503#-N8e59sBBCEypVooG1yU:-N9CRVrr8twnoda95i3e:b-vlqfhw) ([raw file](https://github.com/cockroachdb/cockroach/blob/9230f9aaef000ded75efa63258a183128ae51a75/pkg/kv/kvclient/rangecache/range_cache.go#L387)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

You were planning to address this TODO in this PR, right?
</blockquote></details>

I did address this below but forgot to remove the TODO from here. Fixed. 
___
*[`pkg/kv/kvclient/rangecache/range_cache.go` line 391 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/85503#-N8eQXcpBK3jGP0gpgYm:-N9Hqky649_GM2knxvFn:bon7dsx) ([raw file](https://github.com/cockroachdb/cockroach/blob/9230f9aaef000ded75efa63258a183128ae51a75/pkg/kv/kvclient/rangecache/range_cache.go#L391)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

How should this method behave when this sync updates the leaseholder but the `maybeUpdate` below does not? Should it return `true`?
</blockquote></details>

This relates to your comment above. Given this resembles the old behaviour, and it isn't a correctness issue, how do you feel about keeping things as is for now and then completely getting rid of this return value in a foOne
___
*[`pkg/kv/kvclient/rangecache/range_cache.go` line 410 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/85503#-N8eLqVm-5oBOm8SoSCJ:-N9CTOEW4WgAZne53gZ_:bs7wumx) ([raw file](https://github.com/cockroachdb/cockroach/blob/9230f9aaef000ded75efa63258a183128ae51a75/pkg/kv/kvclient/rangecache/range_cache.go#L410)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

> !descsCompatible(rangeDesc, et.Desc())

We're checking this condition twice. Should we rewrite as:
```go
if !descsCompatible(rangeDesc, et.Desc()) {
    if rangeDesc.Generation < et.Desc().Generation {
        return false
    }
    ...
}
```
</blockquote></details>

Done. 
___
*[`pkg/kv/kvclient/rangecache/range_cache.go` line 426 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/85503#-N8ePyxTAwzJxKR02VEO:-N9HknFlA4n-Jp_RJ60B:bd6iaje) ([raw file](https://github.com/cockroachdb/cockroach/blob/9230f9aaef000ded75efa63258a183128ae51a75/pkg/kv/kvclient/rangecache/range_cache.go#L426)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

Are we ok with returning false here if we updated the descriptor but not the leaseholder? Will the client handle that case correctly?
</blockquote></details>

We should be okay, as the client uses the updated range descriptor to skip any replicas that exist on the transport it constructed but are no longer part of the range descriptor. This happens regardless and is agnostic to whether this function had any effect or not. 
___
*[`pkg/kv/kvclient/rangecache/range_cache.go` line 1338 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/85503#-N8eMEcM2VkUXT_wXXfn:-N9HpX4XDj43y7QWoP3J:bs7wumx) ([raw file](https://github.com/cockroachdb/cockroach/blob/9230f9aaef000ded75efa63258a183128ae51a75/pkg/kv/kvclient/rangecache/range_cache.go#L1338)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

Do you want to give this paragraph a similar "If updatedLease is returned true, ..." sentence?
</blockquote></details>

Done. 
___
*[`pkg/kv/kvclient/rangecache/range_cache.go` line 1340 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/85503#-N8eMfhzA8dNelUWLauL:-N9HpPWE0JT6MgakjVjz:b-lnioii) ([raw file](https://github.com/cockroachdb/cockroach/blob/9230f9aaef000ded75efa63258a183128ae51a75/pkg/kv/kvclient/rangecache/range_cache.go#L1340)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

I found the combination of an `updatedLease` and an `updated` return value to be confusing. It wasn't clear from the caller whether one implied the other or why I needed to care about both. It would be good to detail that.

It's minor, but I think switching the order of these return values would help. I say that because if `updated` is a broader condition than `updatedLease`, then most callers will first consult `updated` and then only care about `updatedLease` is `updated` is true.
</blockquote></details>

Changed the wording in this comment to be a bit more specific and switched the order of these return values. 
___
*[`pkg/kv/kvclient/rangecache/range_cache.go` line 1363 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/85503#-N8eOR_ODe8pT5zv3vnP:-N9DQ6EM5qJmPT2ADfLF:bv2l6dp) ([raw file](https://github.com/cockroachdb/cockroach/blob/9230f9aaef000ded75efa63258a183128ae51a75/pkg/kv/kvclient/rangecache/range_cache.go#L1363)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

I feel like this method would be easier to reason about if you initially cloned the receiver and then updated to lease and desc, instead of working backwards like you're doing here (assuming both will be updated unless proven otherwise). Was there are reason to write it like this?
</blockquote></details>

The reason I wrote it this way was to preserve the behavior tested here: 

https://github.com/cockroachdb/cockroach/blob/5845f4bc637bff4595d9ad9feb5fbaade875559a/pkg/kv/kvclient/rangecache/range_cache_test.go#L1797-L1804

This seems to be related to this comment:

```	// We handle the case of a lease with the sequence equal to the existing
	// entry, but otherwise different. This results in the new lease updating the
	// entry, because the existing lease might correspond to a proposed lease that
	// a replica returned speculatively while a lease acquisition was in progress.
```

However, I don't think this test case is valid, given speculative leases should have 0 sequence number. This also ties in to your comment about `l.Equal(e.Lease()` below. Does this check out?

I'm going to simplify this logic a bit and clone the receiver like you suggest. I'll also delete the test case above. 
___
*[`pkg/kv/kvclient/rangecache/range_cache.go` line 1387 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/85503#-N8eNieJ0voWUDM6afB6:-N9DQxOT1-gdXq2Epio0:bumc7mp) ([raw file](https://github.com/cockroachdb/cockroach/blob/9230f9aaef000ded75efa63258a183128ae51a75/pkg/kv/kvclient/rangecache/range_cache.go#L1387)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

> l.Equal(e.Lease())

I know if was here before, but this is suspect. It's comparing every field in the lease, which I'm not sure we ever want to do on the client. Is it needed to determine that two speculative leases are identical? If not, are there cases where the lease sequence is insufficient to distinguish the equality of leases?
</blockquote></details>

This ties into your comment above. 


<!-- Sent from Reviewable.io -->

nvb reviewed (commented):
> Reviewed 7 of 7 files at r6, 1 of 1 files at r7, 7 of 7 files at r8, 1 of 2 files at r9, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/85503)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @aayushshah15, @arulajmani, and @nvanbenschoten)

___
*[`-- commits` line 36 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/85503#-N8dmtIH5k_kuUAk0_P_):*
<details><summary><i>Previously, arulajmani (Arul Ajmani) wrote…</i></summary><blockquote>

I _think_ the only scenario this is possible is the following scenario:

The cache had no lease information to begin with. It receives a NLHE from a replica that has both a stale range descriptor + lease information. In such a scenario, as written, we choose to insert the lease into the cache. It's possible the stale lease and cached range descriptor are incompatible. 

In my recent round of changes, I chose to structure things such that we don't compare range descriptor generations if the cached lease is empty. The idea being that a stale range descriptor doesn't necessarily imply stale lease information, so in the absence of anything better, let's just try it. How do you feel about this? 
</blockquote></details>

I think this makes sense to me.
___
*[`pkg/kv/kvclient/rangecache/range_cache.go` line 368 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/85503#-N8e5Jyc0DtrjWA1meSa:-N9XCpuHE4Bmw_zSi2qO:baygj01) ([raw file](https://github.com/cockroachdb/cockroach/blob/9230f9aaef000ded75efa63258a183128ae51a75/pkg/kv/kvclient/rangecache/range_cache.go#L368)):*
> Like we discussed offline, much of this can be simplified by changing the conditional at the caller to be lh.IsSame(curReplica) and we can clean up the amount of state that is tracked here and passed between callers. I've added a TODO to this effect, but I'll probably get to it in a quick followup patch if that's okay.

This sounds great. Removing the return value entirely will be a nice simplification. A followup patch SGTM.
___
*[`pkg/kv/kvclient/rangecache/range_cache.go` line 1431 at r8](https://reviewable.io/reviews/cockroachdb/cockroach/85503#-N9XEp1GBdKUMhgB7roR:-N9XEp1GBdKUMhgB7roS:b-akr1ra) ([raw file](https://github.com/cockroachdb/cockroach/blob/8d9b094caa3efd18151fe673700bbcafa48ef07e/pkg/kv/kvclient/rangecache/range_cache.go#L1431)):*
> ```Go
> 	// lease regardless of the associated range descriptor's generation.
> 	if (l.Sequence != 0 && e.lease.Sequence != 0 && l.Sequence > e.lease.Sequence) ||
> 		(l.Sequence == 0 && rangeDesc.Generation >= e.desc.Generation) ||
> ```

nit: I think this can be structured differently to avoid some redundancy and more clearly demonstrate the intent:

```
if ((l.Sequence != 0 && e.lease.Sequence != 0) && l.Sequence > e.lease.Sequence) ||
    ((l.Sequence == 0 || e.lease.Sequence == 0) && rangeDesc.Generation >= e.desc.Generation) ||
    e.lease.Empty() {
```
___
*[`pkg/kv/kvserver/store_send.go` line 196 at r9](https://reviewable.io/reviews/cockroachdb/cockroach/85503#-N9XFpvO4XpL8ZRzzDyn:-N9XFpvO4XpL8ZRzzDyo:b-jcgd4v) ([raw file](https://github.com/cockroachdb/cockroach/blob/21262eb431a013fdd496b36793ecfd53db9ebf92/pkg/kv/kvserver/store_send.go#L196)):*
> ```Go
> 			// tries will be initialized and will have useful leaseholder information
> 			// for the client.
> 			return nil, nil, roachpb.NewError(&roachpb.NotLeaseHolderError{
> ```

I was hoping that this could eliminate the "speculative lease" concept entirely. That doesn't seem to be the case, as there are other places where we return speculative leases. Can we at least unify this with the speculative lease concept? That would mean either removing the LeaseHolder field and returning `Lease: roachpb.Lease{Replica: repl.creatingReplica}` here or eliminating cases where `NotLeaseHolderError.Lease` does not refer to a committed lease and using only the `LeaseHolder` field for speculative leases. Either one seems like a simplification.

Also, since there's more discussion to be had here, consider lifting this into its own PR so that it doesn't block this one.

EDIT: I now remember why we decided to remove this — because the client will ignore the lease due to the missing range desc. My question above still stands regarding what we want to do with these two fields in `NotLeaseHolderError`. This also begs the question of whether we can remove `Replica.creatingReplica`. 


<!-- Sent from Reviewable.io -->

arulajmani reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/85503)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @aayushshah15 and @nvanbenschoten)

___
*[`pkg/kv/kvclient/rangecache/range_cache.go` line 368 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/85503#-N8e5Jyc0DtrjWA1meSa:-N9XmGvdBbTqf84rdmjm:b-qhjy86) ([raw file](https://github.com/cockroachdb/cockroach/blob/9230f9aaef000ded75efa63258a183128ae51a75/pkg/kv/kvclient/rangecache/range_cache.go#L368)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

> Like we discussed offline, much of this can be simplified by changing the conditional at the caller to be lh.IsSame(curReplica) and we can clean up the amount of state that is tracked here and passed between callers. I've added a TODO to this effect, but I'll probably get to it in a quick followup patch if that's okay.

This sounds great. Removing the return value entirely will be a nice simplification. A followup patch SGTM.
</blockquote></details>

Sounds good :+1: 
___
*[`pkg/kv/kvclient/rangecache/range_cache.go` line 1431 at r8](https://reviewable.io/reviews/cockroachdb/cockroach/85503#-N9XEp1GBdKUMhgB7roR:-N9XasCS5dof3HQY3Rlg:bs7wumx) ([raw file](https://github.com/cockroachdb/cockroach/blob/8d9b094caa3efd18151fe673700bbcafa48ef07e/pkg/kv/kvclient/rangecache/range_cache.go#L1431)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

nit: I think this can be structured differently to avoid some redundancy and more clearly demonstrate the intent:

```
if ((l.Sequence != 0 && e.lease.Sequence != 0) && l.Sequence > e.lease.Sequence) ||
    ((l.Sequence == 0 || e.lease.Sequence == 0) && rangeDesc.Generation >= e.desc.Generation) ||
    e.lease.Empty() {
```
</blockquote></details>

Done. 
___
*[`pkg/kv/kvserver/store_send.go` line 196 at r9](https://reviewable.io/reviews/cockroachdb/cockroach/85503#-N9XFpvO4XpL8ZRzzDyn:-N9XnFmr5uM-rOOnblBs:bge31fd) ([raw file](https://github.com/cockroachdb/cockroach/blob/21262eb431a013fdd496b36793ecfd53db9ebf92/pkg/kv/kvserver/store_send.go#L196)):*
> or eliminating cases where NotLeaseHolderError.Lease does not refer to a committed lease and using only the LeaseHolder field for speculative leases. 

Good point on unifying the speculative lease concept -- I like your suggestion about using just the `LeaseHolder` field for speculative leases. I'll open up a followup PR with this change.

For this patch I've removed `Replica.creatingReplica` as this guy was the only place making use of it. 


<!-- Sent from Reviewable.io -->

nvb reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

> Reviewed 6 of 7 files at r12, 4 of 4 files at r13, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/85503#-:-N9YjRHu83rcZOg2xZom:bnfp4nl)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @aayushshah15)



<!-- Sent from Reviewable.io -->

arulajmani commented:
TFTRs! 

bors r+
